### DISCUSSION TITLE
datasource(pypi): simple pypi lookup is not working for packages with normalized names

### DISCUSSION THREAD

**Discussion by Shegox** [renovatebot/renovate#30707] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

38.25.1

### Please tell us more about your question or problem

Renovate, after the v38 update, seems to be no longer able to lookup packages names containing special characters (e.g. `.`) when looking them up in a simple registry (e.g. in a private Artifactory). 

The underlaying problem seems that Renovate is normalizing the name (e.g. replacing `.` with `-`) and only uses the `-`-name to perform the lookup). 

As a concrete example the `ruamel.yaml` package is normalized to `ruamel-yaml` and Renovate then fails to query the datasource.

I think this might be related to https://github.com/renovatebot/renovate/pull/27733

- Reproduction repository (with lookup against `https://www.pypi.org/simple` as this forces Renovate to use the simple lookup method): https://github.com/Shegox/congenial-funicular
- Issue in dependency dashboard: `Renovate failed to look up the following dependencies: Failed to look up pypi package ruamel-yaml.`
- Other non-normalized packages like `requests` work as expected: https://github.com/Shegox/congenial-funicular/pull/4

Notice: Lookups against the standard pypi.org registry use the pypi-API directly and are not affected.

---

I think the problem is within the `extractVersionFromLinkText`-function, which takes the packageName, normalizes it and then uses it for the lookup: 
https://github.com/renovatebot/renovate/blob/6d30cc968033cb62b7e21cc452024ce370c14c0e/lib/modules/datasource/pypi/index.ts#L178-L185

Digging deeper it seems that the `PypiDatasource.normalizeName(text)`-function only normalizes `_` to `-` instead of the additional `.` to `-` as requested by https://packaging.python.org/en/latest/specifications/name-normalization/ : 
https://github.com/renovatebot/renovate/blob/6d30cc968033cb62b7e21cc452024ce370c14c0e/lib/modules/datasource/pypi/index.ts#L85-L87

We do have a correct normalization function in 
https://github.com/renovatebot/renovate/blob/dbd17ed1d78595631b10481e9bf97815fb9acccd/lib/modules/datasource/pypi/common.ts#L9-L12

The problem is however that this will as well normalize a `srcText` like `ruamel.yaml-0.1.tar.gz` to `ruamel-yaml-0-1-tar-gz`. 
We can adjust however our compare logic within the `extractVersionFromLinkText` to do this correctly. 

If there are no objections I will create a PR for this.

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: hostRules: no authentication for www.pypi.org (repository=Shegox/congenial-funicular)
DEBUG: Using queue: host=www.pypi.org, concurrency=16 (repository=Shegox/congenial-funicular)
DEBUG: Failed to look up pypi package ruamel-yaml (repository=Shegox/congenial-funicular, packageFile=requirements.txt, dependency=ruamel-yaml)
DEBUG: PackageFiles.add() - Package file saved for base branch (repository=Shegox/congenial-funicular, baseBranch=main)
DEBUG: Package releases lookups complete (repository=Shegox/congenial-funicular, baseBranch=main)
DEBUG: branchifyUpgrades (repository=Shegox/congenial-funicular)
DEBUG: detectSemanticCommits() (repository=Shegox/congenial-funicular)
DEBUG: getCommitMessages (repository=Shegox/congenial-funicular)
DEBUG: semanticCommits: detected "unknown" (repository=Shegox/congenial-funicular)
DEBUG: semanticCommits: disabled (repository=Shegox/congenial-funicular)
DEBUG: 1 flattened updates found: requests (repository=Shegox/congenial-funicular)
DEBUG: Returning 1 branch(es) (repository=Shegox/congenial-funicular)
DEBUG: config.repoIsOnboarded=true (repository=Shegox/congenial-funicular)
DEBUG: packageFiles with updates (repository=Shegox/congenial-funicular, baseBranch=main)
       "config": {
         "pip_requirements": [
           {
             "deps": [
               {
                 "depName": "ruamel.yaml",
                 "packageName": "ruamel-yaml",
                 "currentValue": "==0.18.5",
                 "datasource": "pypi",
                 "currentVersion": "0.18.5",
                 "updates": [],
                 "versioning": "pep440",
                 "warnings": [
                   {
                     "topic": "ruamel-yaml",
                     "message": "Failed to look up pypi package ruamel-yaml"
                   }
                 ]
               },

```

</details>

**Comment by rarkins** [comment]
Yes please go ahead!

**Comment by rarkins** [comment, marked as answer]
Converted to issue: https://github.com/renovatebot/renovate/issues/30712
